src/Kernel/Magic/DebugMagic.cs

```diff
@@ -0,0 +1,246 @@
+                ManualResetEvent? @event = null;
+                lock (sessionAdvanceEvents)
+                {
+                    @event = sessionAdvanceEvents[sessionGuid];
+                }
+                @event.Set();
+            }
+        });
+
+        private async Task WaitForAdvance(Guid session)
+        {
+            await Task.Run(() =>
+            {
+                ManualResetEvent? @event = null;
+                // Find the event we need to wait on.
+                lock (sessionAdvanceEvents)
+                {
+                    @event = sessionAdvanceEvents[session];
+                }
+                @event.Reset();
+                @event.WaitOne();
+            }, cancellationTokenSource.Token);
+        }
+
+        /// <summary>
+        ///     Simulates an operation given a string with its name and a JSON
+        ///     encoding of its arguments.
+        /// </summary>
+        public async Task<ExecutionResult> RunAsync(string input, IChannel channel)
+
+        {
+            var inputParameters = ParseInputParameters(input, firstParameterInferredName: ParameterNameOperationName);
+
+            var name = inputParameters.DecodeParameter<string>(ParameterNameOperationName);
+            var symbol = SymbolResolver.Resolve(name) as IQSharpSymbol;
+            if (symbol == null) throw new InvalidOperationException($"Invalid operation name: {name}");
+
+            var session = Guid.NewGuid();
+            var divId = session.ToString();
+            lock (sessionAdvanceEvents)
+            {
+                sessionAdvanceEvents[session] = new ManualResetEvent(true);
+            }
+
+            using var qsim = new QuantumSimulator();
+            qsim.OnOperationStart += (callable, args) =>
+            {
+                // Tell the IOPub channel that we're starting a new operation.
+                shellServer.SendIoPubMessage(
+                    new Message
+                    {
+                        Header = new MessageHeader
+                        {
+                            MessageType = "iqsharp_debug_opstart"
+                        },
+                        Content = new DebugStatusContent
+                        {
+                            DebugSession = session.ToString(),
+                            State = new DisplayableState
+                            {
+
+                                QubitIds = qsim.QubitIds.Select(q => (int)q),
+                                NQubits = (int) (qsim.QubitManager?.GetAllocatedQubitsCount() ?? 0),
+                                Amplitudes = new DebugStateDumper(qsim).getAmplitudes(),
+                                DivId = divId
+                            }
+                        }
+                    }
+                );
+                WaitForAdvance(session).Wait();
+            };
+
+            channel.Display(new RawHtmlPayload
+            {
+                //needs to be changed?
```

Comment:
This comment could probably be removed, unless there's something left to change here?